### ISSUE TITLE
Tags "tokens" accessibility evaluation

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#1339]
Splitting this out from #1141

When adding tags to a post, the added tags are displayed as sort of "labels" before the input field:

<img width="374" alt="screen shot 2017-06-21 at 17 18 41" src="https://user-images.githubusercontent.com/1682452/27392024-c8526234-56a5-11e7-8aae-431372ebd980.png">

Visually, they appear to be inside the input field but they're actually _before_ the field, which I think is a good thing.

The rendered markup of a tag "label" is going to be improved a bit in #1141 and it's something like this:
```
<span class="components-form-token-field__token" tabindex="-1">
	<span class="components-form-token-field__token-text">audio</span>
	<button type="button" aria-label="Remove tag: audio." class="components-button components-icon-button components-form-token-field__remove-token">
		<svg aria-hidden="true" role="img" ... ></svg>
	</button>
</span>
```

They're just span elements with no particular semantics, not wrapped in any list or other semantic element (same in the current WP implementation). From an accessibility perspective, worth considering a few improvements:
- currently, they're not operable with a keyboard: once the field has focus, it is not possible to tab backwards to them: this is a keyboard trap that should be fixed
- should these tags be wrapper in a list?
- is it OK they're _before_ the field? IN teh current WP implementation, they're _after_ the field and this helps discoverability. Placing them before the field could make finding them a bit difficult for screen reader users and make keyboard interaction not so ideal (users are forced to tab backwards)

Aside: from an usability perspective, I'd propose to put the "X" on the left: this way, removing tags would be easier because it would be possible to click always in the same place to remove multiple tags. Click, click, click, the X is always in the same place.

**Comment by afercia** [comment]
Looking a bit at this, there's a keyboard trap now:
- add a tag
- focus the field
- press Shift+Tab to move away from the field
- nothing happens, the field is still focused

Additionally, when the field is focused, the attribute `aria-expanded` is always true: this should happen only when the list of suggestions is rendered.

Seems to me the `isActive` state should be reviewed, it can't be true when the input is focused.

**Comment by afercia** [comment]
Also, compared to the current behavior in the classic editor, pressing Tab when there are no highlighted suggestions should not add a tag. To reproduce:
- go in the classic editor
- in the tag field, type `xyz`
- press Tab
- no tag gets added, and this is correct
- do the same in Gutenberg
- type `xyz` and press Tab
- a new `xyz` tag gets created and added

This behavior is dangerous, as there's the risk of creating dozens of unwanted tags. I'd recommend to consider to add a button to explicitly add a new tag.

**Comment by afercia** [comment]
Same for the comma handling: it might seem a smart idea to create a new tag when typing a comma, but it's probably not. For example, if I type a new tag and then a comma, the tag is created and added but what if I've made a typo and the newly added tag was not the intended one/ Or, I could just inadvertently press the comma key and get a new tag added. Soon, my list of tags would be filled up with lots of unwanted tags.

**Comment by afercia** [comment]
I'm seeing a small glitch where the caret is not visible in the input field, especially on Firefox. To reproduce:
- add 2 tags then Tab away from the input field
- Shift+Tab to move back to the input field
- the "real" input field is focused, as expected, but the caret is not visible
- start typing something (even just a space) and the caret is visible again

**Comment by aardrian** [comment]
This is strictly a response to the code sample at the start of the issue. I wanted to address what @afercia raised and give some context for those reading in a vacuum. Very simplified code suggestion:

```
<li …>
	<span … id="Tag01">audio</span>
	<button -type="button"- aria-label="Remove tag." … aria-describedby="Tag01">
		<svg aria-hidden="true" role="img" ... ></svg>
	</button>
</li>
```

I got rid of the classes just to simplify. I put the individual tag in a `<li>`. For the `<button>` I suggest removing `type=button`. I simplified the `aria-label` to just "Remove tag." I also added an `id` to the `<span>` wrapping the text label and reference it in the `<button>` with `aria-describedby`.

The idea here is that the list conveys to a screen reader user how many tags there are and the user's current position within the set of tags.

The label text is simplified since a user is ostensibly hearing the button immediately after the text. For users who are just jumping into the button from elsewhere or who forgot the current tag, `aria-describedby` still announces the tag text after a moment. Long enough to keep from annoying a user familiar with the page, but still be useful for those who are not.

**Comment by afercia** [comment]
I'd like to make it a list but I forgot how much this component is complicated 😬  

It has been built in a way (not saying I necessarily agree) that using the arrows allows to move the caret through the added tags:

<img width="295" alt="screen shot 2018-04-13 at 18 14 57" src="https://user-images.githubusercontent.com/1682452/38746427-b85a6a62-3f47-11e8-946f-a014c45c9714.png">

Actually, this is a visual trick: where the caret appears, there's actually an input field that gets moved in the DOM. Using the left arrow it's moved up in the DOM, using the right arrow, it's moved down in the DOM. This allows, using a few events and JS, to emulate the `Backspace` and `Canc` keys behaviors. Once the caret is moved in a position, it is possible to remove the tag placed on the left or the one placed on the right.

Changing the tags in a list would make this behavior impossible. That would require to refactor the whole component and simplify it, removing the emulation trick.  I'm not sure the team leads would agree 🙂 

Maybe we could add some visually hidden text to indicate the number of tags.

**Comment by afercia** [comment]
Playing a bit with this, we can't change it in a list but I think we can add the position and count information this way (simplified code):

```
<span id="someid-1" aria-label="audio (1 of 2)">
	audio
</span>
<button ... aria-describedby="someid-1" aria-label="Remove Tag"></button>
```

Would that be acceptable? @aardrian thanks!

**Comment by afercia** [comment]
Hm maybe I'm going to use a `screen-reader-text` instead of an `aria-label`.

**Comment by afercia** [comment]
What's left to do here, quoting from #6214 

In a following PR I'd like to try to:
- not create/add a new term when tabbing away from the input field: just using the keyboard on this component for a while, I've not intentionally created tons of new terms in my tags list... this should be avoided
- maybe (to evaluate) add the total count of terms added in a visible text before the field, something along the lines of `{plural term name} added: 3`
- ideally, the tokens should be placed after the input field but it's not possible to change this because of the way this component works 😞
